Libraries/LibWeb/DOM/Node.cpp

```diff
@@ -1162,6 +1176,190 @@ WebIDL::ExceptionOr<GC::Ref<Node>> Node::clone_node(Document* document, bool sub
+
+        // NOTE: If we didn't have a layout node before, rebuilding the layout tree isn't gonna give us one
+        //       after we've been removed from the DOM.
+        if (layout_node())
+            old_parent->set_needs_layout_tree_update(true);
+    }
+
+    // 13. Remove node from oldParent’s children.
+    old_parent->remove_child_impl(*this);
+
+    // 14. If node is assigned, then run assign slottables for node’s assigned slot.
+    if (auto assigned_slot = assigned_slot_for_node(*this))
+        assign_slottables(*assigned_slot);
+
+    // 15. If oldParent’s root is a shadow root, and oldParent is a slot whose assigned nodes is empty, then run signal a slot change for oldParent.
+    auto& old_parent_root = old_parent->root();
+    if (old_parent_root.is_shadow_root() && is<HTML::HTMLSlotElement>(*old_parent)) {
+        auto& old_parent_slot = static_cast<HTML::HTMLSlotElement&>(*old_parent);
+        if (old_parent_slot.assigned_nodes_internal().is_empty())
+            signal_a_slot_change(old_parent_slot);
+    }
+
+    // 16. If node has an inclusive descendant that is a slot:
+    auto has_descendent_slot = false;
+
+    for_each_in_inclusive_subtree_of_type<HTML::HTMLSlotElement>([&](auto const&) {
+        has_descendent_slot = true;
+        return TraversalDecision::Break;
+    });
+
+    if (has_descendent_slot) {
+        // 1. Run assign slottables for a tree with oldParent’s root.
+        assign_slottables_for_a_tree(old_parent_root);
+
+        // 2. Run assign slottables for a tree with node.
+        assign_slottables_for_a_tree(*this);
+    }
+
+    // 17. If child is non-null:
+    if (child) {
+        // 1. For each live range whose start node is newParent and start offset is greater than child’s index, increase its start offset by 1.
+        for (auto& range : Range::live_ranges()) {
+            if (range->start_container() == &new_parent && range->start_offset() > child->index())
+                range->increase_start_offset({}, 1);
+        }
+
+        // 2. For each live range whose end node is newParent and end offset is greater than child’s index, increase its end offset by 1.
+        for (auto& range : Range::live_ranges()) {
+            if (range->end_container() == &new_parent && range->end_offset() > child->index())
+                range->increase_end_offset({}, 1);
+        }
+    }
+
+    // 18. Let newPreviousSibling be child’s previous sibling if child is non-null, and newParent’s last child otherwise.
+    auto* new_previous_sibling = child ? child->previous_sibling() : new_parent.last_child();
+
+    // 19. If child is null, then append node to newParent’s children.
+    if (!child) {
+        new_parent.append_child_impl(*this);
+    }
+    // 20. Otherwise, insert node into newParent’s children before child’s index.
+    else {
+        new_parent.insert_before_impl(*this, child);
+    }
+
+    new_parent.invalidate_style(StyleInvalidationReason::NodeInsertBefore);
```

Comment:
I'm a little unsure about how correct the invalidation is here. @kalenikaliaksandr  maybe you have some insight into how this should be done?

I think we need to be doing invalidation on both `new_parent` (here) after insertion into the tree and `old_parent` after removal.

Reply:
invalidating both new and old parents sounds correct to me

Reply:
iirc the way `invalidate_style` is called has changed recently due to Andreas hitting the performance go brrr button. This should be revistied